framework/include/messaging/messaging.h

```diff
@@ -0,0 +1,159 @@
+/**
+ * @brief The type of receiving message
+ * @details MSG_RECV_TYPE_BLOCK    : A receiver blocks until receiving the message.\n
+ *	    MSG_RECV_TYPE_NONBLOCK : A receiver sets callback function for receiving message and continues to run.
+ */
+enum msg_recv_type_e {
+	MSG_RECV_TYPE_BLOCK = 1,
+	MSG_RECV_TYPE_NONBLOCK = 2,
+	MSG_RECV_TYPE_MAX
+};
+typedef enum msg_recv_type_e msg_recv_type_t;
+
+/**
+ * @brief The structure of data buffer which used for receiving
+ */
+struct msg_recv_data_s {
+	char *msg;
+	int msglen;
+};
+typedef struct msg_recv_data_s msg_recv_data_t;
+
+/**
+ * @brief The structure of parameters which used for receiving
+ */
+struct msg_recv_param_s {
+	msg_recv_type_t type; /* for receiver : MSG_RECV_BLOCK or MSG_RECV_NONBLOCK */
+	msg_callback_t cb_func;
+	char *cb_data;
+};
+typedef struct msg_recv_param_s msg_recv_param_t;
+/**
+ * @brief Send unicast message to specified message port
+ * @details @b #include <messaging/messaging.h>\n
+ * @param[in] port_name The message port name to send
+ * @param[in] send_param
+ *		  msg          : The message to be sent\n
+ *		  msglen       : The length of message to be sent\n
+ *		  priority     : A non-negative integer that specifies the priority of this message\n
+ *		  type         : MSG_SEND_SYNC or MSG_SEND_ASYNC\n
+ * @param reply_param
+ *		  [out] reply_msg    : A message buffer to receive the sync reply message\n
+ *		  [in] reply_msglen  : A message size for sync reply message
+ * @return On success, OK is returned. On failure, Error is returned.
+ * @since TizenRT v3.0
+ */
+int messaging_send(const char *port_name, msg_send_param_t *send_param, msg_reply_param_t *reply_param);
+/**
+ * @brief Wait to receive unicast message from specified message port
+ * @details @b #include <messaging/messaging.h>\n
+ * @param[in] port_name The message port name to receive
+ * @param data\n
+ *		[out] msg     : The message buffer to receive the message\n
+ *		[in] msglen  : The length of message to receive\n
+ * @param[in] param\n
+ *		type    : MSG_RECV_BLOCK or MSG_RECV_NONBLOCK\n
+ *		cb_func : user callback function for nonblocking receive\n
+ *		cb_data : user callback data for nonblocking receive\n
+ * @return On success, OK is returned. On failure, Error is returned.
+ * @since TizenRT v3.0
+ */
+int messaging_recv(const char *port_name, msg_recv_data_t *data, msg_recv_param_t *param);
+/**
+ * @brief Reply unicast message to specified message port when received sync message
+ * @details @b #include <messaging/messaging.h>\n
+ * @param[in] port_name The message port name to send\n
+ *		This port name should be matched to original port name which received the message.
+ * @param[in] reply_param
+ *		  reply_msg    : The message to be sent\n
+ *		  reply_msglen : The length of message to be sent
+ * @return On success, OK is returned. On failure, Error is returned.
+ * @since TizenRT v3.0
+ */
+int messaging_send_reply(const char *port_name, msg_reply_param_t *reply_param);
```

Comment:
How to know when we should use this? Is there any way that receiver knows sync type?